tensorflow/core/ops/batchnorm_training_op.cc

```diff
@@ -0,0 +1,80 @@
+/* Copyright 2015 Google Inc. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/framework/op.h"
+
+namespace tensorflow {
+
+REGISTER_OP("BatchNormalizeTraining")
```

Comment:
It is okay to have a separate BatchNormalizeInference op for now. However, in the future, we really want a single "BatchNorm" that can choose "inference" and "training" variants depending on whether gradients are needed. That will be an improvement down the road. 

Regarding the name, TF calls it BatchNorm. Both Cudnn and the original paper use the term BatchNormalization. I would prefer not adding a new variant, either one works for me.